Specify Custom Fields

NobodysNightmare opened the pull request:
## OpenProject Work Packages
- remotely related to https://community.openproject.org/work_packages/14051
## Progress
- [x] Specify schema as own resource
- [x] Specify endpoint for work package schema
- [x] Extend schema specification to suffice for reading purposes and custom fields
- [x] Extend work package specification on how additional (non core) fields are represented
- [x] specify endpoint for list custom field values (i.e. an "identity transform" endpoint)
## Changes to existing implementation
- form includes link to schema resource
- schema resource promoted to a real type (with own endpoint, self-link, etc.)
- schema now needs to include more details and those for all the properties

NobodysNightmare committed:
move schema out into its own endpoint

[ci skip]

TeatroIO commented:
I've prepared a stage. <a href='http://feature-specify-custom-fields.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Click to open.</a>

NobodysNightmare committed:
add some more properties to field schema

[ci skip]

NobodysNightmare committed:
remove schema link from form again

linking to the right schema is not as easy as I thought ^^

[ci skip]

NobodysNightmare committed:
define remaining schema properties

[ci skip]

NobodysNightmare committed:
reference schema endpoint for work packages

[ci skip]

NobodysNightmare commented:
This specification contains an inconspicuous detail. The URL for the schema of a work package is defined as `/api/v3/work_packages/schemas/{identifier}` where `{identifier}` is a **string**.

In practice this means the implementation has a single string to reference all possible schemas a work package can have.

The naive approach is to let the work package be the identifier of the schema, thereby having one unique schema URL per work package. However, this approach will not employ caching mechanics very good, as multiple work packages that share the same schema will refer to their common schema by different URLs.

The schema (for reading purposes) of a work package (as of today) depends on the following factors:
- the project of the work package
- the type of the work package

In the future more factors might get into that equation.

Thus, ideally a schema is referenced using these distinguishing factors only, so that multiple work packages sharing the same factors provide a link to the same URL.

One approach to this could be to encode all relevant factors as first class resources in the URL, e.g.

```
/api/v3/projects/{project_id}/work_packages/schemas/types/{type_id}

e.g.
/api/v3/projects/12/work_packages/schemas/types/13
```

This has some downsides:
- The URL is cluttered with details and not really understandable in itself (why does the type come after the schema?
- The URL exposes implementation details (the schema of a WP depends on project and type)

But at least there is the advantage of good cachability.

Another approach is the one currently chosen and favorited by me:

```
/api/v3/work_packages/schemas/{identifier}

e.g.
/api/v3/work_packages/schemas/12-13
```

What is that? The "identifier" happens to contain all relevant information to refer to a schema.

This has a downside too: This Identifier is rather arbitrary and the client may not understand it.

Personally I see this as a feature:
- a HAL client should only follow URLs and not try to make up his own URLs, so it does not matter if it understands the meaning of "12-13" other than being a random identifier
- Implementation details are a little more consealed, it _should_ be easier to introduce more factors contributing to the schema later on (this would simply change the identifier, but not the whole URL structure)

## tl; dr:

The URL `/api/v3/work_packages/schemas/12-13` encodes information that could be made more obvious into an identifier, mainly for performance considerations. Do you think this is a good idea or a bad idea? (considering alternatives)

NobodysNightmare committed:
extend examples

[ci skip]

NobodysNightmare committed:
Merge remote-tracking branch 'origin/dev' into feature/specify_custom_fields

Conflicts:
	doc/apiv3-documentation.apib

[ci skip]

ulferts commented on doc/apiv3-documentation.apib:
```diff
@@ -1510,6 +1501,127 @@ but are also limited to the projects that the current user is allowed to see.
+| minLength         | The value of the property must at least contain the specified amount of characters | Integer  | 0       |
+| maxLength         | The value of the property must at most contain the specified amount of characters  | Integer  | ∞       |
+| regularExpression | The value of the property must match the given regular expression (if not null)    | String   | null    |
+| required          | If true this property might not be null                                            | Boolean  | true    |
```
"might not be null" is weird wording IMO.

ulferts commented on doc/apiv3-documentation.apib:
```diff
@@ -1510,6 +1501,127 @@ but are also limited to the projects that the current user is allowed to see.
+| maxLength         | The value of the property must at most contain the specified amount of characters  | Integer  | ∞       |
+| regularExpression | The value of the property must match the given regular expression (if not null)    | String   | null    |
+| required          | If true this property might not be null                                            | Boolean  | true    |
+| visible           | If true this property can be displayed in a UI                                     | Boolean  | true    |
```
@NobodysNightmare and I talked about this and agreed, that it is the client's responsibility to decide what to display.

ulferts commented:
@NobodysNightmare the allowedValues for list custom fields still need definition.

NobodysNightmare committed:
more concise wording

NobodysNightmare committed:
remove visible property from schema

a client should decide by itself which fields to display and which not

[ci skip]

NobodysNightmare committed:
specify string object endpoint

[ci skip]

NobodysNightmare committed:
fix typos, examples and add notes

[ci skip]

ulferts commented:
LGTM. In case anybody want's to give some more feedback (@myabc, @linki, @machisuji), I am giving this a grace period till the end of today, before I merge it.
